docs/proposals/202105256-cluster-class-and-managed-topologies.md

```diff
@@ -0,0 +1,480 @@
+  // of the template to an infrastructure cluster.
+  Infrastructure LocalObjectTemplate `json:"infrastructure,omitempty"`
+}
+
+// WorkerClass is a collection of deployment classes which pronounce
+//
+type WorkerClass struct {
+  // Deployments is a list of machine deployment types that can be used to create
+  // a set of worker plane nodes.
+  Deployments []DeploymentClass `json:"deployments,omitempty"`
+}
+
+
+// DeploymentClass serves as a template to define a set of worker nodes of the cluster
+// provisioned using the `ClusterClass`.
+type DeploymentClass struct {
+  // Class denotes a type of worker node present in the cluster
+  Class string `json:"class,omitempty"`
+
+
+  // Template is a local struct containing a collection of templates for creation of
+  // MachineDeployment objects representing a set of worker nodes.
+  Template TemplateClass `json:"template"`
+}
+
+type TemplateClass struct {
+  Metadata ObjectMeta `json:"metadata,omitempty"`
+
+  // Bootstrap contains the bootstrap template references to be used
+  // for the creation of worker Machines.
+  Bootstrap LocalObjectTemplate `json:"bootstrap,omitempty"`
+
+  // Infrastructure contains the infrastructure template references to be used
+  // for the creation of worker Machines.
+  Infrastructure LocalObjectTemplate `json:"infrastructure,omitempty"`
+}
+
+type LocalObjectTemplate struct {
+  // Ref is a required reference to a custom resource
+  // offered by a bootstrap provider.
+  Ref corev1.ObjectReference `json:"ref"`
+}
+```
+
+#### Modification to existing API Types
+##### Cluster
+1.  Add `Cluster.Spec.Topology` defined as
+    ```golang
+    // This encapsulates the managed topology for the cluster.
+    // +optional
+    Topology *ManagedTopology `json:"topology,omitempty"`
+    ```
+1.  The `ManagedTopology` object has the following definition:
+    ```golang
+    // ManagedTopology encapsulates the information of the managed resources.
+    type ManagedTopology struct {
+      // The name of the ClusterClass object to create the managed topology.
+      Class string `json:"class,omitempty"`
+
+      // The kubernetes version of the cluster.
+      Version string `json:"version"`
+
+      // RolloutAfter performs a rollout of the entire cluster one component at a time [...]
+      // +optional
+      RolloutAfter *metav1.Time `json:"rolloutAfter,omitempty"`
+
+      // The information for the Control plane of the cluster.
+      ControlPlane ControlPlane `json:"controlPlane"`
+
+      // Worker encapsulates the different constructs that form the worker nodes
+      // for the cluster.
+      // +optional
+      Worker *Worker `json:"worker,omitempty"`
+    }
+    ```
+1.  The `ControlPlane` object contains the parameters for the control plane nodes of the managed topology.
+    ```golang
+    // ControlPlane specifies the parameters for the control plane nodes in the cluster.
+    type ControlPlane struct {
+     Metadata ObjectMeta `json:"metadata,omitempty"`
```

Comment:
Does a full ObjectMeta makes sense here or only a subset? (similar in other cases like Worker)

Reply:
This `ObjectMeta` struct is the CAPI struct defined in `v1a4/common_types.go`. It just contains the `Labels` and `Annotations` both of which are being used. I think you have it confused with the `ObjectMeta` from api-machinery. Hopefully, that resolves it.